src/pages/single-product/single-product.tsx

```diff
@@ -0,0 +1,95 @@
+import { useContext, useState } from "react";
+import { useNavigate, useParams } from "react-router-dom";
+import { FALLBACK_IMAGE, PRODUCTS_ENDPOINT } from "../../constants";
+import { CartContextObject } from "../../entities/cart-context-object";
+import useProduct from "../../hooks/use-product";
+import { CartContext } from "../../providers/cart-provider";
+import { ProductContext } from "../../providers/products-provider";
+import APIClient from "../../services/api-client";
+import CartService from "../../services/cart-service";
+
+const SingleProduct = () => {
+  const { id } = useParams();
+  const navigate = useNavigate();
+
+  const { data: product, error, isLoading } = useProduct(id!);
+  const [quantity, setQuantity] = useState(1);
+  const { cart, setCart } = useContext<CartContextObject>(CartContext);
+  const { contextProducts, refetchProducts } = useContext(ProductContext);
+
+  const handleQuantityChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const newQuantity = parseInt(event.target.value, 10);
+    setQuantity(newQuantity);
+  };
+
+  const handleDelete = async () => {
+    if (confirm(`Are you sure you want to delete ${product?.name}?`)) {
+      await new APIClient(PRODUCTS_ENDPOINT).remove(id);
+      navigate("/products");
+      const newProducts = contextProducts.filter((item) => item.id !== id);
+      refetchProducts(newProducts);
+      const cartItem = cart.find((item) => item.product.id === id)!;
```

Comment:
I honestly don't like the non-null assertion (id!) and I would rather do a check before.
Check this out: https://blog.italanta.net/the-case-against-non-null-assertion-operator-ea6848aed636

Reply:
Yes, indeed. In this scenario I wrongly assumed that a product that I am deleting is in the cart as well. Here a conditional statement is almost mandatory if using common sense.

However, I think there are scenarios where the non-null assertion may be useful, such as this case where I am retrieving the product id using `useParams()` hook inside the SingleProduct component, that is routed to a path containing an id.
As I cannot call hooks conditionally I must somehow guarantee to the compiler that the id is not null. (I am using `useEffect()` inside `useProduct()`.

![image](https://github.com/user-attachments/assets/bc79dd30-c3a3-4a44-987a-227131337fcd)

P.S.: Really cool article, thank you.